CMakeLists.txt

```diff
@@ -472,38 +486,137 @@ function(_display_cpp_implementation_msg name)
 endfunction()
-if (BLAS++)
-  _display_cpp_implementation_msg("BLAS")
-  include(ExternalProject)
-  ExternalProject_Add(blaspp
-	  URL https://bitbucket.org/icl/blaspp/downloads/blaspp-2020.10.02.tar.gz
-	  CONFIGURE_COMMAND ${CMAKE_COMMAND} -E env LIBRARY_PATH=$ENV{LIBRARY_PATH}:${CMAKE_BINARY_DIR}/lib LD_LIBRARY_PATH=$ENV{LD_LIBRARY_PATH}:${PROJECT_BINARY_DIR}/lib ${CMAKE_COMMAND} -DCMAKE_INSTALL_PREFIX=${PROJECT_BINARY_DIR} -DCMAKE_INSTALL_LIBDIR=lib -DBUILD_SHARED_LIBS=${BUILD_SHARED_LIBS} ${PROJECT_BINARY_DIR}/blaspp-prefix/src/blaspp
-	  BUILD_COMMAND ${CMAKE_COMMAND} -E env LIBRARY_PATH=$ENV{LIBRARY_PATH}:${PROJECT_BINARY_DIR}/lib LIB_SUFFIX="" ${CMAKE_COMMAND} --build .
-	  INSTALL_COMMAND ${CMAKE_COMMAND} -E env PREFIX=${PROJECT_BINARY_DIR} LIB_SUFFIX="" ${CMAKE_COMMAND} --install .
-	  )
-  ExternalProject_Add_StepDependencies(blaspp build ${BLAS_LIBRARIES})
+
+if (BLAS++ OR LAPACK++)
+  
+  if (BLAS++)
+    _display_cpp_implementation_msg("BLAS")
+  endif()
+
+  # Check if population has already been performed
+  FetchContent_GetProperties(blaspp)
+  if(NOT blaspp_POPULATED)
+    # Fetch the content using previously declared details
+    FetchContent_Populate(blaspp)
+  endif()
+
+  # For static builds, we may need to link against a Fortran library
+  set(BLAS_Fortran_LIB "")
+  if (NOT BLAS_FOUND AND NOT BUILD_SHARED_LIBS)
+    if (CMAKE_Fortran_COMPILER_ID MATCHES GNU)
+      set(BLAS_Fortran_LIB ";-lgfortran")
+    else()
+      # TODO: This is incomplete. Fill in the other cases.
+      set(BLAS_Fortran_LIB "")
+    endif()
+  endif()
+
+  # Adds target blaspp
+  add_custom_target( blaspp ALL DEPENDS blaspp-cmd )
+  add_custom_command( OUTPUT blaspp-cmd
+    WORKING_DIRECTORY "${blaspp_SOURCE_DIR}"
+    COMMENT "Building BLAS++" )
+
+  # Set up information about the BLAS and LAPACK libraries
+  if(NOT BLAS_FOUND)
+    if(NOT LATESTLAPACK_FOUND)
+      add_custom_command( OUTPUT blaspp-cmd APPEND
+        COMMAND ${CMAKE_COMMAND}
+        -B "${blaspp_BINARY_DIR}"
+        -D BLAS_LIBRARIES="$<TARGET_FILE:${BLASLIB}>${BLAS_Fortran_LIB}"
+        -D LAPACK_LIBRARIES="$<TARGET_FILE:${LAPACKLIB}>" )
+    else()
+      add_custom_command( OUTPUT blaspp-cmd APPEND
+        COMMAND ${CMAKE_COMMAND}
+        -B "${blaspp_BINARY_DIR}"
+        -D BLAS_LIBRARIES="$<TARGET_FILE:${BLASLIB}>${BLAS_Fortran_LIB}"
+        -D LAPACK_LIBRARIES="${LAPACK_LIBRARIES}" )
+    endif()
+  else()
+    if(NOT LATESTLAPACK_FOUND)
+      add_custom_command( OUTPUT blaspp-cmd APPEND
+        COMMAND ${CMAKE_COMMAND}
+        -B "${blaspp_BINARY_DIR}"
+        -D BLAS_LIBRARIES="${BLAS_LIBRARIES}"
+        -D LAPACK_LIBRARIES="$<TARGET_FILE:${LAPACKLIB}>${BLAS_Fortran_LIB}" )
+    else()
+      add_custom_command( OUTPUT blaspp-cmd APPEND
+        COMMAND ${CMAKE_COMMAND}
+        -B "${blaspp_BINARY_DIR}"
+        -D BLAS_LIBRARIES="${BLAS_LIBRARIES}"
+        -D LAPACK_LIBRARIES="${LAPACK_LIBRARIES}" )
+    endif()
+  endif()
```

Comment:
These 4 `add_custom_command` blocks are quite redundant. May I suggest setting some variables to have one 1 instance? See my top-level review comment for suggestion.

Reply:
The redundancy is because we cannot use CMake generator expressions (e.g., `$<TARGET_FILE:${LAPACKLIB}>`) to set CMake variables at configuration time. See https://stackoverflow.com/questions/51353110/how-do-i-output-the-result-of-a-generator-expression-in-cmake. There is certainly a better way to do this, but I couldn't figure it out.